### PR TITLE
order budget history cleanup

### PR DESCRIPTION
**Pull request by julianhyatt**
- add cleanup command for cleanup of old order budget history entries

### CODE REVIEW

File: bundles/order-budget/src/FondOfOryx/Shared/OrderBudget/OrderBudgetConstants.php [hunk 1]
```diff
@@ -13,4 +13,14 @@ interface OrderBudgetConstants
...
+
+    /**
+     * @var string
+     */
+    public const RETENTION_TIME_IN_DAYS = 'FOND_OF_ORYX:ORDER_BUDGET:RETENTION_TIME_IN_DAYS';
+
+    /**
+     * @var int
+     */
+    public const RETENTION_TIME_IN_DAYS_DEFAULT = 360;
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Shared/OrderBudget/OrderBudgetConstants.php]: Why 360? 365 = 1 year

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace FondOfOryx\Zed\OrderBudget\Business\Cleanupper;
+
+use DateTime;
+use FondOfOryx\Zed\OrderBudget\OrderBudgetConfig;
+use FondOfOryx\Zed\OrderBudget\Persistence\OrderBudgetEntityManagerInterface;
+use Spryker\Zed\Kernel\Persistence\EntityManager\TransactionTrait;
+
+class OrderBudgetHistoryCleanupper implements OrderBudgetHistoryCleanupperInterface
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php]:
```suggestion
class OrderBudgetHistoryCleaner implements OrderBudgetHistoryCleanerInterface
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        OrderBudgetConfig $config
+    ) {
+        $this->entityManager = $entityManager;
+        $this->config = $config;
+    }
+
+    /**
+     * @return void
+     */
+    public function removeOldHistoryEntries(): void
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php]:
```suggestion
    public function clean(): void
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        });
+    }
+
+    /**
+     * @return void
+     */
+    protected function executeCleanup(): void
+    {
+        $maxDate = date('Y-m-d', strtotime(sprintf('- %s days', $this->config->getHistoryRetentionTime())));
+        $test = new DateTime($maxDate);
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleanupper/OrderBudgetHistoryCleanupper.php]:
```suggestion
        $oldestDate = (new DateTime())->sub(new DateInterval(sprintf('P%sD', $this->config->getHistoryRetentionTime())));
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/OrderBudgetFacade.php [hunk 5]
```diff
@@ -26,6 +26,19 @@ public function resetOrderBudgets(array $orderBudgetIds = []): void
...
 
+    /**
+     * Specification:
+     * - Removes old data from history
+     *
+     * @api
+     *
+     * @return void
+     */
+    public function removeOldOrderBudgetsFromHistory(): void
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/OrderBudgetFacade.php]:
```suggestion
    public function cleanupOrderBudgetHistory(): void
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * @method \FondOfOryx\Zed\OrderBudget\Persistence\OrderBudgetRepositoryInterface getRepository()
+ * @method \FondOfOryx\Zed\OrderBudget\Business\OrderBudgetFacadeInterface getFacade()
+ */
+class OrderBudgetHistoryCleanupConsole extends Console
+{
+    /**
+     * @var string
+     */
+    public const COMMAND_NAME = 'order-budget:history-cleanup';
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php]:
```suggestion
    public const COMMAND_NAME = 'order-budget:history:cleanup';
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @var string
+     */
+    public const DESCRIPTION = 'This command will remove old order budgets.';
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php]:
```suggestion
    public const DESCRIPTION = 'This command will remove old history entries of order budgets.';
```

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     *
+     * @return int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->info('Order budgets history cleanup');
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Communication/Console/OrderBudgetHistoryCleanupConsole.php]:
```suggestion
        $this->info('Cleanup order budget history');
```

**Review by daniel-rose — CHANGES_REQUESTED** [review]

File: bundles/order-budget/src/FondOfOryx/Shared/OrderBudget/OrderBudgetConstants.php (hunk 1 above)
Comment (julianhyatt) [reply bundles/order-budget/src/FondOfOryx/Shared/OrderBudget/OrderBudgetConstants.php]: a bank year is 360 days

File: bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleaner/OrderBudgetHistoryCleaner.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+        $this->getTransactionHandler()->handleTransaction(static function () use ($self) {
+            $self->executeCleanup();
+        });
+    }
+
+    /**
+     * @return void
+     */
+    protected function executeCleanup(): void
```
Comment (daniel-rose) [inline bundles/order-budget/src/FondOfOryx/Zed/OrderBudget/Business/Cleaner/OrderBudgetHistoryCleaner.php]:
```suggestion
    protected function executeClean(): void
```

**Review by daniel-rose — APPROVED** [review]

**Pull request merged**